common/src/main/java/com/microsoft/identity/common/internal/controllers/BrokerOperationExecutor.java

```diff
@@ -0,0 +1,187 @@
+ * - If the current strategy fails to connect to the broker, it will try the next one until the list is exhausted.
+ */
+public class BrokerOperationExecutor {
+
+    private static final String TAG = BrokerOperationExecutor.class.getSimpleName();
+
+    /**
+     * Info of a broker operation to be performed with available strategies.
+     */
+    public interface BrokerOperation<T> {
+
+        /**
+         * Trigger any prerequisite works before making the actual request.
+         * This was added because the existing MSAL-Broker logic has a separate hello() call.
+         */
+        void performPrerequisites(@NonNull IIpcStrategy strategy) throws BaseException;
+
+        /**
+         * Gets a BrokerOperationBundle bundle to pass to each IpcStrategies.
+         */
+        @NonNull BrokerOperationBundle getBundle();
+
+        /**
+         * Extracts the result object from a bundle returned by an IpcStrategy.
+         * If the broker returns an error, this will throw an exception.
+         */
+        @NonNull T extractResultBundle(@Nullable final Bundle resultBundle) throws BaseException;
+
+        /**
+         * Returns method name (for logging/telemetry purpose).
+         */
+        @NonNull String getMethodName();
+
+        /**
+         * ID of the telemetry API event associated to this strategy task.
+         * If this value returns null, no telemetry event will be emitted.
+         */
+        @Nullable String getTelemetryApiId();
+
+        /**
+         * A method that will be invoked before the success event is emitted.
+         * If the calling operation wants to put any value in the success event, put it here.
+         */
+        void putValueInSuccessEvent(ApiEndEvent event, T result);
+    }
+
+    private final List<IIpcStrategy> mStrategies;
+
+    /**
+     * @param strategies list of IIpcStrategy to be invoked.
+     */
+    public BrokerOperationExecutor(@NonNull final List<IIpcStrategy> strategies) {
+        mStrategies = strategies;
+    }
+
+    /**
+     * A generic method that would initialize and iterate through available strategies.
+     * It will return a result immediately if any of the strategy succeeds, or throw an exception if all of the strategies fails.
+     */
+    public <T extends CommandParameters, U> U execute(@NonNull final T parameters,
+                                                      @NonNull final BrokerOperation<U> operation)
+            throws BaseException {
+
+        if (operation.getTelemetryApiId() != null) {
+            Telemetry.emit(
+                    new ApiStartEvent()
+                            .putProperties(parameters)
+                            .putApiId(operation.getTelemetryApiId())
+            );
+        }
+
+        U result = null;
+        Exception lastCaughtException = null;
+        for (int ii = 0; ii < mStrategies.size(); ii++) {
+            final IIpcStrategy strategy = mStrategies.get(ii);
```

Comment:
Here we can time the strategy execution from outside, by using the getName() method which, for bonus points, is by contract the telemetry name.

Reply:
I'll make this as part of the ests telemetry work.